### PR TITLE
SHIP-0028: BuildRun cleanup

### PR DESCRIPTION
**Pull request by SaschaSchwarze0**
Looking for some feedback already especially on the TBDs inside the ship.

### CODE REVIEW

File: ships/0028-buildrun-cleanup.md
Comment (gabemontero) [inline ships/0028-buildrun-cleanup.md:81]: My assumption is that if *either* the TTL or limit count is reached, pruning occurs.

But let's not assume it is obvious (which ever way you are thinking) and state that here, even if you make it clearer in the backend/details section.

By comparison, I would expect that distinction to be clear in the godoc of the API go files vs. having to pick up on that fact in the associated markdown files.

**Review by gabemontero — COMMENTED** [review]
a couple of quick thoughts on what you have @SaschaSchwarze0

File: ships/0028-buildrun-cleanup.md
Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md]: That's also matching my gut feeling. Added it to the agenda for next Monday to discuss it. https://github.com/shipwright-io/community/issues/62#issuecomment-1041378397

Comment (gabemontero) [reply ships/0028-buildrun-cleanup.md]: sounds good

File: ships/0028-buildrun-cleanup.md
Comment (qu1queee) [inline ships/0028-buildrun-cleanup.md:68]: on the wording, `successful` vs `succeeded` . Succeeded inlines with how we do in the `.status.conditions`, see https://github.com/shipwright-io/build/blob/main/docs/buildrun.md#understanding-the-state-of-a-buildrun

File: ships/0028-buildrun-cleanup.md
Comment (qu1queee) [inline ships/0028-buildrun-cleanup.md:100]: any reasons why you didnt choose the `lasttransitiontime` of our `Succeeded` condition. I believe is more accurate as it based on our controller updates(last state transition)

File: ships/0028-buildrun-cleanup.md
Comment (qu1queee) [inline ships/0028-buildrun-cleanup.md:136]: +1, this contradicts the one-shot previous efforts, due to extra reconcile loops.

File: ships/0028-buildrun-cleanup.md
Comment (qu1queee) [inline ships/0028-buildrun-cleanup.md:142]: Can we consider having a separate controller? I need to spend more time thinking on the pros/cons of this.

**Review by qu1queee — CHANGES_REQUESTED** [review]

File: ships/0028-buildrun-cleanup.md
Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md:100]: Makes sense, had forgotten that we take `completionTime` from the TaskRun.

File: ships/0028-buildrun-cleanup.md
Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md:142]: Yep, I had that idea as well. Need to think about this again now that we discussed how we handle Build updates.

File: ships/0028-buildrun-cleanup.md
Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md]: Done.

File: ships/0028-buildrun-cleanup.md
Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md]: Done.

File: ships/0028-buildrun-cleanup.md
Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md]: Done.

File: ships/0028-buildrun-cleanup.md
Comment (qu1queee) [reply ships/0028-buildrun-cleanup.md:142]: ok, let me try to brainstorm today/tomorrow, before putting my thoughts here. Thanks.

Comment (qu1queee) [reply ships/0028-buildrun-cleanup.md:142]: Adding here my thoughts(already expressed them offline to Sascha):

My concern on not using another controller, is in terms of the reconcile loop increase. My rational in terms of reconciles is the following(based on the worst scenario on ttl):

Assuming 600 completed Buildruns with TTL. 
- Today a single BR reconciles 4-5 times only. => 3000 reconciles
- We increase per BR +1 reconcile(due to the requeue)  => +600 reconciles
- On rollouts of the controller,  assuming all 600 are completed and waiting for a requeue(worst scenario), thats a +600 (we didnt do this before) => +600 reconciles
- In total +1200, a 40% plus. Significant increase.

An alternative for the rollouts would be introducing rate limiting in the controller, see [ratelimiter](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/ratelimiter#RateLimiter) that might help with one of the extra 600.

Comment (gabemontero) [reply ships/0028-buildrun-cleanup.md:142]: Some quick "prior art" for reference.  OpenShift build pruning has handled things in "the same" controller / reconciler loop for many years with little to no problems, but OpenShift builds does not have TTL based pruning, only count/total based.

It will be interesting to see as we work the implementation which of the following approaches turn out most palatable:
- count based pruning in the existing controller/reconciler loop, but ttl based in a new controller/reconciler
- both count and ttl based in a new controller/reconciler
- both count and ttl based in the existing controller/reconciler

Seeing this also resulted in the following notion popping into my head:  do we need both TTL and count based pruning out of the gate?  i.e. is there some CodeEngine specific req wanting TTL?  Or is it sufficient to go out with say just count based pruning initially, then wait for feedback on whether that is sufficient?

FWIW, there have been no requests for TTL based pruning that I can recall from OpenShift Build customers (where OCP Builds was first released in 2015).

Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md:142]: Thanks @gabemontero for the feedback. The main motivation to have the TTL is related to the standalone BuildRuns where there is no Build (see the other ship from Baran). If there is no Build, there is no count-based pruning.

**Comment by SaschaSchwarze0** [comment]
@qu1queee @gabemontero I added a commit with updates. Those mainly cover two things:

1. Describing the behavior when Build updates happen after BuildRuns were completed.
2. I changed the implementation considerations to use new controllers for the cleanup. Knowing the complexity of the BuildRun reconciler, I feel much better with this approach as it would have meant many additional `if`-style code in the existing one and the increased complexity would have further increased the risk that we oversee certain scenarios.

Please have a look. Thank you.

File: ships/0028-buildrun-cleanup.md
Comment (qu1queee) [inline ships/0028-buildrun-cleanup.md:101]: to be really precise 

> Condition `type=Succeeded` changes its `.Status` field, from ...

File: ships/0028-buildrun-cleanup.md
Comment (qu1queee) [inline ships/0028-buildrun-cleanup.md:151]: yes, we need this.

**Review by qu1queee — APPROVED** [review]
Approving. Looks neat, moving forward with two separate controllers is ideal, if we would have opt-int for extending the current controllers only, then we might have just started with the limits, and not ttl.

/approve

File: ships/0028-buildrun-cleanup.md
Comment (gabemontero) [reply ships/0028-buildrun-cleanup.md:157]: ah thanks for the clarification @SaschaSchwarze0 ... yeah, I have not had a chance to look at that EP yet, and the title of `One-Off Builds` did not immediately connect with me that it meant "BuildRuns without Builds", but yes it makes sense, that without that organizational aspect of "being under a Build", the count based pruning does not make sense.  Or more precisely, the only option is "global" limits and that is not what is desired.

I'm wrapping up my review of the totality of the latest changes now.  Thanks.

**Review by gabemontero — APPROVED** [review]
I've got nothing to add @SaschaSchwarze0 .  I see @qu1queee added a couple of comments and while I don't *think* he is asking for changes (instead just making anecdotal comments) I'm not 100% sure, so I'll refrain on adding a lgtm label, along with I see you have @adambkaplan listed as an approver, so I'm taking that as you wanting him to take a pass at this as well before it merges.

File: ships/0028-buildrun-cleanup.md
Comment (SaschaSchwarze0) [reply ships/0028-buildrun-cleanup.md:112]: :-)

**Comment by SaschaSchwarze0** [comment]
> I've got nothing to add @SaschaSchwarze0 . I see @qu1queee added a couple of comments and while I don't _think_ he is asking for changes (instead just making anecdotal comments) I'm not 100% sure, so I'll refrain on adding a lgtm label, along with I see you have @adambkaplan listed as an approver, so I'm taking that as you wanting him to take a pass at this as well before it merges.

I added the remaining small changes (status of conditions as mentioned by Enrique, and the SHOP vs SHIP typo found by Raghav). I also change the approvers section. I only initialized them that way. Commits also squashed. Looking for a lgtm. @gabemontero ;-)

File: ships/0028-buildrun-cleanup.md
Comment (adambkaplan) [reply ships/0028-buildrun-cleanup.md:80]: +1 to using `Duration`.

I also like having separate TTLs for success and failure - a failed BuildRun may need a longer TTL so the dev team can inspect logs.

File: ships/0028-buildrun-cleanup.md
Comment (adambkaplan) [inline ships/0028-buildrun-cleanup.md:104]: +1

File: ships/0028-buildrun-cleanup.md
Comment (adambkaplan) [inline ships/0028-buildrun-cleanup.md:113]: +1 to this as well - standalone `BuildRun` objects can be considered "orphans" and should only use TTL for cleanup.

File: ships/0028-buildrun-cleanup.md
Comment (adambkaplan) [inline ships/0028-buildrun-cleanup.md:152]: Perhaps we enforce an upper bound on succeeded + failed limits in the main `Build` controller and/or an additional webhook? Something to consider as a future enhancement.

File: ships/0028-buildrun-cleanup.md
Comment (adambkaplan) [inline ships/0028-buildrun-cleanup.md:156]: More of a nit - the drawbacks are addressed in the "Risks and Mitigations" section.

**Review by adambkaplan — APPROVED** [review]
/lgtm

This is a great proposal @SaschaSchwarze0! I've noted a few follow up areas in the proposal that we may want to consider. Otherwise this looks good.

**Pull request merged**
